modules/indexer/code/indexer.go

```diff
@@ -185,7 +186,7 @@ func Init() {
 
 			rIndexer, populate, err = NewBleveIndexer(setting.Indexer.RepoPath)
 			if err != nil {
-				if rIndexer != nil {
+				if !giteautil.IsInterfaceNil(rIndexer) {
```

Comment:
I think the function `NewBleveIndexer` is doing wrong.

If an error is returned, then the `rIndexer` should have been destroyed inside the `NewBleveIndexer`, caller shouldn't do more clean up work. Current `NewBleveIndexer` breaks the Go standard.

Reply:
> If an error is returned, then the `rIndexer` should have been destroyed inside the `NewBleveIndexer`, caller shouldn't do more clean up work. Current `NewBleveIndexer` breaks the Go standard.

That's interesting, that is indeed correct. Should that be a upstream-issue?

I think we can remove the closing code - as it's totally unnecessary as you say. 

> And I find that `NewBleveIndexer` always returns a `&BleveIndexer{}`, why do we need `IsInterfaceNil` here?

We don't actually need it as it's unnecessary then.

Reply:
Should we also patch `NewBleveIndexer`?

```
func NewBleveIndexer(indexDir string) (*BleveIndexer, bool, error) {
	indexer := &BleveIndexer{
		indexDir: indexDir,
	}
	if created, err := indexer.init(); err != nil {
		indexer.Close()
		return nil, false, err
	}
	return indexer, created, nil
}
```